includes/class-debug.php

```diff
@@ -58,29 +66,97 @@ public static function log( $message, $type = 'string' ) {
+	 *
+	 * @return string|WP_Error The File content truncate upto the number of lines set in the limit parameter.
+	 */
+	public static function read( $limit = 1000 ) {
+		$wp_filesystem = self::init_filesystem();
+		$file_path     = self::get_file_path();
+		if ( ! self::verify_filesystem( $wp_filesystem ) || ! $wp_filesystem->exists( $file_path ) || ! $wp_filesystem->is_readable( $file_path ) ) {
+			return new \WP_Error( 'not_readable', __( 'Error: Unable to read the log file.', 'AspireUpdate' ) );
+		}
 
-		$file_path = WP_CONTENT_DIR . '/' . self::$log_file;
+		$file_content = $wp_filesystem->get_contents_array( $file_path );
+		$content      = '';
+		$index        = 0;
+		foreach ( $file_content as $file_content_lines ) {
+			if ( ( $index < $limit ) ) {
+				$content .= $file_content_lines . PHP_EOL;
+				++$index;
+			}
+		}
+		if ( '' === trim( $content ) ) {
+			$content = esc_html__( '*****Log file is empty.*****', 'AspireUpdate' );
+		} elseif ( $limit < count( $file_content ) ) {
+			$content .= PHP_EOL . sprintf(
+				/* translators: 1: The number of lines at which the content was truncated. */
+				esc_html__( '*****Log truncated at %s lines.*****', 'AspireUpdate' ),
+				$limit
+			);
+		}
+		return $content;
+	}
 
-		$content = '';
-		if ( $wp_filesystem->exists( $file_path ) ) {
-			$content = $wp_filesystem->get_contents( $file_path );
+	/**
+	 * Clear content of the log file.
+	 *
+	 * @return boolean|WP_Error true on success and false on failure.
+	 */
+	public static function clear() {
+		$wp_filesystem = self::init_filesystem();
+		$file_path     = self::get_file_path();
+		if ( ! self::verify_filesystem( $wp_filesystem ) || ! $wp_filesystem->exists( $file_path ) || ! $wp_filesystem->is_writable( $file_path ) ) {
+			return new \WP_Error( 'not_accessible', __( 'Error: Unable to access the log file.', 'AspireUpdate' ) );
 		}
 
 		$wp_filesystem->put_contents(
 			$file_path,
-			$content . $formatted_message,
+			'',
 			FS_CHMOD_FILE
 		);
+		return true;
+	}
+
+	/**
+	 * Logs a message to the debug log file.
+	 *
+	 * @param mixed  $message The message to log.
+	 * @param string $type   The log level ('string', 'request', 'response').
+	 */
+	public static function log( $message, $type = 'string' ) {
+		$wp_filesystem = self::init_filesystem();
+		if ( self::verify_filesystem( $wp_filesystem ) ) {
+			$timestamp         = gmdate( 'Y-m-d H:i:s' );
+			$formatted_message = sprintf(
+				'[%s] [%s]: %s',
+				$timestamp,
+				strtoupper( $type ),
+				self::format_message( $message )
+			) . PHP_EOL;
+
+			$file_path = self::get_file_path();
+
+			$content = '';
+			if ( $wp_filesystem->exists( $file_path ) ) {
+				if ( $wp_filesystem->is_readable( $file_path ) ) {
+					$content = $wp_filesystem->get_contents( $file_path );
+				}
+			}
+			$wp_filesystem->put_contents(
+				$file_path,
+				$formatted_message . $content,
+				FS_CHMOD_FILE
+			);
+		}
 	}
```

Comment:
Aside from being incredibly inefficient by rewriting the entire file every time, this is inherently a race condition: if two requests come in at the same time and the first one finishes after the second, whatever the second request logged will be clobbered by the first.  The WP filesystem API has no capability to open a file for appending, so we need to just not use it for this.

Reply:
These changes are from `main`